plugins/clnrest/utilities/rpc_routes.py

```diff
@@ -0,0 +1,71 @@
+import json5
+from flask import request, make_response, Response, stream_with_context
+from flask_restx import Namespace, Resource
+from .shared import call_rpc_method, verify_rune, process_help_response
+from .rpc_plugin import plugin
+
+methods_list = []
+rpcns = Namespace("RPCs")
+payload_model = rpcns.model("Payload", {}, None, False)
+
+@rpcns.route("/list-methods")
+class ListMethodsResource(Resource):
+    @rpcns.response(200, "Success")
+    @rpcns.response(500, "Server error")
+    def get(self):
+        """Get the list of all valid rpc methods"""
+        try:
+            help_response = call_rpc_method(plugin, "help", [])
+            html_content = process_help_response(help_response)
+            response = make_response(html_content)
+            response.headers["Content-Type"] = "text/html"
+            return response
+
+        except Exception as err:
+            plugin.log(f"Error: {err}", "error")
+            return json5.loads(str(err)), 500
+
+@rpcns.route("/<rpc_method>")
+class RpcMethodResource(Resource):
+    @rpcns.doc(security=[{"rune": [], "nodeid": []}])
+    @rpcns.doc(params={"rpc_method": (f"Name of the RPC method to be called")})
+    @rpcns.expect(payload_model, validate=False)
+    @rpcns.response(201, "Success")
+    @rpcns.response(500, "Server error")
+    def post(self, rpc_method):
+        """Call any valid core lightning method (check list-methods response)"""
+        try:
+            is_valid_rune = verify_rune(plugin, request)
+            
+            if "error" in is_valid_rune:
+                plugin.log(f"Error: {is_valid_rune}", "error")
+                raise Exception(is_valid_rune)
+
+        except Exception as err:
+            # Fix Me: Remove after lightningd checkrune is available
+            if not "unknown command" in str(err).lower():
+                return json5.loads(str(err)), 403
+        
+        try:
+            if request.is_json:
+                payload = request.get_json()
+            else:
+                payload = request.form.to_dict()
+            return call_rpc_method(plugin, rpc_method, payload), 201
+
+        except Exception as err:
+            plugin.log(f"Error: {err}", "error")
+            return json5.loads(str(err)), 500
+
+@rpcns.route("/notifications")
+class NotificationsResource(Resource):
+    def get(self):
+        try:
+            def notifications_stream():
+                while True:
+                    from .rpc_plugin import queue
+                    yield queue.get()
```

Comment:
Can this queue.get() return events that are inserted while looping? So this will send event inconsistently back!

we can do `for _ in queue.qsize()` and return only the notification that are inside the queue at the time of the call

Reply:
The bigger issue is that this is not a broadcast queue, but rather a single-producer-multi-consumer one that fans out. This means that if we have two clients calling the `/notifications` endpoint they'll get alternating events, not a full copy of the events as they happen.

It'd be better to have a high-watermark that you pass in with the query, and the server returns immediately if it has the `n+1`st event ready, or block if there is no `n+1`st. This means we no longer have watchers racing, rather we have a very controlled fetch next behavior.

For garbage collection we could just drop anything that is older than a given age, or something that has not been retrieved for some time.

Reply:
@cdecker Thank you for pointing it out. I somehow missed the obvious drawback here. I will implement it with Websockets now.